Update string-interner dependency to newest

Robbepop opened the pull request:
Version `0.11` of the `string-interner` crate significantly reduces overall memory consumption and increases internment throughput.

Link to crate: https://crates.io/crates/string-interner

For `string-interner` version `0.12` it is planned to add yet another pluggable backend that is inspired directly by `strena`. Btw.: The plotter always just plots "THIS LIB" instead of the library name.

Looking at the stats it seems that `string-interner` actually uses even less memory than `lasso`. :partying_face: 

```
     total events | 66
      peak bytes  | 319.9 KB
     ----------------------------
     alloc events | 41
     alloc bytes  | 492.3 KB
     ----------------------------
     freed events | 25
     freed bytes  | 213.4 KB
```

![2020-07-14-195847_2117x1132_scrot](https://user-images.githubusercontent.com/8193155/87460023-7a2e8e00-c60c-11ea-8896-cefe331b29ca.png)

CAD97 commented:
> Btw.: The plotter always just plots "THIS LIB" instead of the library name.

That would be because all of the scripts here are super sketch and I've just been manually setting the plot title and yticks 😅

> Also benchmarks show that [hashbrown] is 20-30% faster than Rust's hashmap implementation.

Rust's std uses hashbrown as the hashmap, so any speedup would be due to more aggressive inlining from hashbrown by default (the `inline-more` feature). It'd be interesting to see comparisons between hashbrown with and without `inline-more`, and without `inline-more` but with `lto=fat` (along with compile time comparisons). When hashbrown was brought to std it was found that turning off the (newly added for this purpose) inline-more feature gave a huge compile speed improvement for hashmap-heavy crates (like cargo) for minimal performance penalty. (The reason: `#[inline]` inlines into every CGU in addition to the other hinting it does, so you end up compiling the hash table multiple times in your one crate.)

Robbepop commented:
> > Btw.: The plotter always just plots "THIS LIB" instead of the library name.
> 
> That would be because all of the scripts here are super sketch and I've just been manually setting the plot title and yticks sweat_smile
> 
> > Also benchmarks show that [hashbrown] is 20-30% faster than Rust's hashmap implementation.
> 
> Rust's std uses hashbrown as the hashmap, so any speedup would be due to more aggressive inlining from hashbrown by default (the `inline-more` feature). It'd be interesting to see comparisons between hashbrown with and without `inline-more`, and without `inline-more` but with `lto=fat` (along with compile time comparisons). When hashbrown was brought to std it was found that turning off the (newly added for this purpose) inline-more feature gave a huge compile speed improvement for hashmap-heavy crates (like cargo) for minimal performance penalty. (The reason: `#[inline]` inlines into every CGU in addition to the other hinting it does, so you end up compiling the hash table multiple times in your one crate.)

Thanks for letting me know! So I guess adding a crate feature to control this `inline-more` as done in `strena` should be added to `string-interner` as well.

CAD97 commented:
Results from my machine:

string-interner:

     total events | 66
      peak bytes  | 319.9 KB
     ----------------------------
     alloc events | 41
     alloc bytes  | 492.3 KB
     ----------------------------
     freed events | 25
     freed bytes  | 213.4 KB

lasso:

     total events | 41
      peak bytes  | 409.7 KB
     ----------------------------
     alloc events | 24
     alloc bytes  | 634.0 KB
     ----------------------------
     freed events | 17
     freed bytes  | 285.8 KB

The difference is probably due to arena growth strategies, if I had to guess. I'd have to analyze the data a bit more closely to know for sure. Lasso finishes the test with fewer alloc events but by requesting a bit more total memory.

Robbepop commented:
> Results from my machine:
> 
> string-interner:
> 
> ```
>  total events | 66
>   peak bytes  | 319.9 KB
>  ----------------------------
>  alloc events | 41
>  alloc bytes  | 492.3 KB
>  ----------------------------
>  freed events | 25
>  freed bytes  | 213.4 KB
> ```
> 
> lasso:
> 
> ```
>  total events | 41
>   peak bytes  | 409.7 KB
>  ----------------------------
>  alloc events | 24
>  alloc bytes  | 634.0 KB
>  ----------------------------
>  freed events | 17
>  freed bytes  | 285.8 KB
> ```
> 
> The difference is probably due to arena growth strategies, if I had to guess. I'd have to analyze the data a bit more closely to know for sure.

Yep would make sense (I thought the same) because `string-interner` has more allocation/free events. So I'd say it is a trade-off.

btw.: What exactly did you mean here?

> (To be completely fair: the indirect map does have a measurable impact on throughput, so the extra denormalization does have a purpose and isn't wasted space.)

Positive or negative effect? From my own measurements string internment efficiency actually improved alongside memory consumption. So it was win-win if I didn't make a mistake with my benchmarks.

CAD97 commented:
> From my own measurements string internment efficiency actually improved alongside memory consumption. So it was win-win if I didn't make a mistake with my benchmarks.

I'm surprised! I did a _very_ rough quick experiment that suggested that the extra indirection involved in the indirect map caused lookups to be slower. Given you didn't see that effect, I expect for an actual workload that the cache coherency benefits dominate the small cost of the extra indirection.

CAD97 commented:
Updated the postscript one more time to demonstrate string-interner v0.11. Additionally, added a graph comparing the two libraries:

![](https://res.cloudinary.com/practicaldev/image/fetch/s--Q6uM5gNh--/c_limit%2Cf_auto%2Cfl_progressive%2Cq_auto%2Cw_880/https://dev-to-uploads.s3.amazonaws.com/i/7sd7lz99vpdi9g3wjsws.png)

It's cool to see the different arena strategies at work here!

Robbepop commented:
> Updated the postscript one more time to demonstrate string-interner v0.11. Additionally, added a graph comparing the two libraries:
> 
> ![](https://camo.githubusercontent.com/fbe9fb216d2ce141f15da6205185c4fd897d8365/68747470733a2f2f7265732e636c6f7564696e6172792e636f6d2f70726163746963616c6465762f696d6167652f66657463682f732d2d5136754d35674e682d2d2f635f6c696d6974253243665f6175746f253243666c5f70726f6772657373697665253243715f6175746f253243775f3838302f68747470733a2f2f6465762d746f2d75706c6f6164732e73332e616d617a6f6e6177732e636f6d2f692f377364376c7a393976706469396733776a7377732e706e67)
> 
> It's cool to see the different arena strategies at work here!

Thanks for this graph - it is absolutely beautiful to look at! :)
Also thanks for updating the blog post, yet again. :see_no_evil:

Robbepop commented:
@CAD97 Update on the `string-interner`:
I just implemented the `strena` backend called `StringBackend` and added some further optimizations towards memory consumption for it. Link to PR: https://github.com/Robbepop/string-interner/pull/26
The results are impressive!

How I was able to get even better memory consumption:
Basically instead of spans I just store the `end` part because the `start` (or the `end`) is redundantly stored.

## `StringBackend`

```
     total events | 75
      peak bytes  | 221.2 KB
     ----------------------------
     alloc events | 39
     alloc bytes  | 360.6 KB
     ----------------------------
     freed events | 36
     freed bytes  | 180.4 KB
```

## Your `strena`

```
total events | 75
  peak bytes  | 254.0 KB
 ----------------------------
 alloc events | 39
 alloc bytes  | 426.1 KB
 ----------------------------
 freed events | 36
 freed bytes  | 213.1 KB
```

![2020-07-15-023702_2137x1164_scrot](https://user-images.githubusercontent.com/8193155/87489818-440d0080-c644-11ea-96af-884b8a6c8f59.png)

I modified your `string-interner-comparison` crate slightly to test it since the new `StringBackend` has not yet been released:

```rust
fn interner_collect_words(&self) {
    crate::ALLOCATOR.set_active(true);
    use string_interner::{
        StringInterner,
        DefaultSymbol,
        DefaultHashBuilder,
        backend::StringBackend,
    };
    let mut words = <StringInterner<DefaultSymbol, StringBackend<DefaultSymbol>, DefaultHashBuilder>>::new();
    crate::ALLOCATOR.mark_point();
    for &word in WORDS {
        words.get_or_intern(word);
        crate::ALLOCATOR.mark_point();
    }
    crate::ALLOCATOR.set_active(false);
    println!("Loaded {} words", words.len());
}
```

Btw.: You wanted some compile-time comparison between `inline-more` and without:

```
> cargo clean
> cargo build --no-default-features --features std
   Compiling autocfg v1.0.0
   Compiling ahash v0.3.8
   Compiling cfg-if v0.1.10
   Compiling hashbrown v0.8.0
   Compiling string-interner v0.11.3 (/home/projects/string-interner)
    Finished dev [unoptimized + debuginfo] target(s) in 1.43s
> cargo clean
> cargo build
   Compiling autocfg v1.0.0
   Compiling serde v1.0.114
   Compiling ahash v0.3.8
   Compiling cfg-if v0.1.10
   Compiling hashbrown v0.8.0
   Compiling string-interner v0.11.3 (/home/projects/string-interner)
    Finished dev [unoptimized + debuginfo] target(s) in 5.35s
```
